[Feature][VectorStore] Support StarRocks as vector db

dirtysalt opened the pull request:
<!--
Thank you for contributing to LangChain! Your PR will appear in our release under the title you set. Please make sure it highlights your valuable contribution.

Replace this with a description of the change, the issue it fixes (if applicable), and relevant context. List any dependencies required for this change.

After you're done, someone will review your PR. They may suggest improvements. If no one reviews your PR within a few days, feel free to @-mention the same people again, as notifications can get lost.

Finally, we'd love to show appreciation for your contribution - if you'd like us to shout you out on Twitter, please also include your handle!
-->

<!-- Remove if not applicable -->

Fixes # (issue)

#### Before submitting

<!-- If you're adding a new integration, please include:

1. a test for the integration - favor unit tests that does not rely on network access.
2. an example notebook showing its use


See contribution guidelines for more information on how to write tests, lint
etc:

https://github.com/hwchase17/langchain/blob/master/.github/CONTRIBUTING.md
-->

Here are some examples to use StarRocks as vectordb

```
from langchain.vectorstores import StarRocks
from langchain.vectorstores.starrocks import StarRocksSettings

embeddings = OpenAIEmbeddings()

# conifgure starrocks settings
settings = StarRocksSettings()
settings.port = 41003
settings.host = '127.0.0.1'
settings.username = 'root'
settings.password = ''
settings.database = 'zya'

# to fill new embeddings
docsearch = StarRocks.from_documents(split_docs, embeddings, config = settings)   


# or to use already-built embeddings in database.
docsearch = StarRocks(embeddings, settings)
```

#### Who can review?

Tag maintainers/contributors who might be interested:

@dev2049 

<!-- For a quicker response, figure out the right person to tag with @

  @hwchase17 - project lead

  Tracing / Callbacks
  - @agola11

  Async
  - @agola11

  DataLoaders
  - @eyurtsev

  Models
  - @hwchase17
  - @agola11

  Agents / Tools / Toolkits
  - @hwchase17

  VectorStores / Retrievers / Memory
  - @dev2049

 -->

dirtysalt committed:
use normalized version `cosine_similarity_norm`

hwchase17 commented on langchain/vectorstores/starrocks.py:
```diff
@@ -0,0 +1,447 @@
+            config: Optional[StarRocksSettings] = None,
+            **kwargs: Any,
+    ) -> None:
+        """StarRocks Wrapper to LangChain
```
better docstring would be nice

dirtysalt replied:
I've updated  docstring and added a example notebook to use starrocks as vectordb.

hwchase17 reviewed (commented):
an example notebook would be very helpful!

dirtysalt commented:
> an example notebook would be very helpful!

OK. I'll fix that.

dirtysalt committed:
fix conflict

rlancemartin commented:
> > an example notebook would be very helpful!
> 
> 
> 
> OK. I'll fix that.

Thanks! I'll have a look at the notebook tomorrow, and help get this merged.

dirtysalt committed:
add more comment

dirtysalt committed:
add example of using starrocks as vectordb

dirtysalt committed:
update example notebook

dirtysalt commented:
> > > an example notebook would be very helpful!
> > 
> > 
> > OK. I'll fix that.
> 
> Thanks! I'll have a look at the notebook tomorrow, and help get this merged.

Thanks. I've added a example notebook to use starrocks as vectordb.

rlancemartin commented:
@dirtysalt thanks!

(1) In the notebook, please add a header `# Starrocks` and a short description with context on the DB. It will help with adoption. E.g., see example for other vectorDBs [here](https://github.com/rlancemartin/langchain/blob/master/docs/extras/modules/data_connection/vectorstores/integrations/deeplake.ipynb).

(2) Please run `make format` to format and also have a look at the lint errors in the tests: 
```
langchain/vectorstores/starrocks.py:26: error: Function is missing a type annotation  [no-untyped-def]
langchain/vectorstores/starrocks.py:32: error: Function is missing a type annotation  [no-untyped-def]
langchain/vectorstores/starrocks.py:33: error: Library stubs not installed for "pymysql.cursors"  [import]
langchain/vectorstores/starrocks.py:33: note: Hint: "python3 -m pip install types-PyMySQL"
langchain/vectorstores/starrocks.py:33: note: (or run "mypy --install-types" to install all missing stub packages)
langchain/vectorstores/starrocks.py:33: note: See https://mypy.readthedocs.io/en/stable/running_mypy.html#missing-imports
langchain/vectorstores/starrocks.py:33: error: Library stubs not installed for "pymysql"  [import]
langchain/vectorstores/starrocks.py:34: note: By default the bodies of untyped functions are not checked, consider using --check-untyped-defs  [annotation-unchecked]
langchain/vectorstores/starrocks.py:[19](https://github.com/hwchase17/langchain/actions/runs/5309392803/jobs/9644405342?pr=6119#step:6:20)3: error: "Iterable[str]" has no attribute "index"  [attr-defined]
langchain/vectorstores/starrocks.py:335: error: Need type annotation for "settings_strs" (hint: "settings_strs: List[<type>] = ...")  [var-annotated]
langchain/vectorstores/starrocks.py:449: error: "StarRocks" has no attribute "client"  [attr-defined]
```

dirtysalt committed:
to format python code

dirtysalt committed:
fix for lint

dirtysalt committed:
update notebook and add some comments

dirtysalt committed:
update notebook and add some comments

dirtysalt committed:
update notebook and add some comments

dirtysalt committed:
update notebook and add some comments

dirtysalt commented:
> 

@rlancemartin I'm quite a newbie to this community, and thanks for pointing out problems.

I've fixed lint and format problems.

And I've updated the notebook and added some descriptions about the background of StarRocks.

Really appreciate your review and time.

dirtysalt committed:
update notebook and add some comments

rlancemartin commented:
> > 
> 
> @rlancemartin I'm quite a newbie to this community, and thanks for pointing out problems.
> 
> I've fixed lint and format problems.
> 
> And I've updated the notebook and added some descriptions about the background of StarRocks.
> 
> Really appreciate your review and time.

Thanks! No problem. Lint errors are always a bit annoying :) I kicked off tests and will look again tomorrow. I can resolve any remaining ones quickly and get this in.

dirtysalt commented:
OK. 

I've checked lint error. Looks like you have to install a python package in lint python env.    

Because I've used `pymysql` this package, and there is no `pymysql` stub for type annotation checking.

> python3 -m pip install types-PyMySQL

dev2049 committed:
ignore

dev2049 committed:
mv nb

dev2049 committed:
nb nit

dirtysalt commented on langchain/vectorstores/starrocks.py:
```diff
@@ -0,0 +1,460 @@
+    def __init__(
+        self,
+        embedding: Embeddings,
+        config: Optional[StarRocksSettings] = None,
```
The assert statement is right after handling the null case

```
        if config is not None:
            self.config = config
        else:
            self.config = StarRocksSettings()
        assert self.config
```

if `config` is None, we give it a default setting.

So in theory, we can remove `assert self.config` this stmt, because it will always true.

dirtysalt committed:
fix typo
